### ISSUE TITLE
UNITY_SUPPORT_64 Core Dumped on 32 bit processor (proposal patch)

### ISSUE THREAD

**Issue by alef75** [ThrowTheSwitch/Unity#343]
Enabling UNITY_SUPPORT_64 and using the 64 bit asserts (e.g. TEST_ASSERT_EQUAL_UINT64)
make the testrunner to coredump.
The problem is in src/unity_internals.h
the UNITY_TEST_ASSERT_XXXX_XX64 sets of macros use UNITY_INTcast  in place of UNITY_INT64.
In attach the patch file to fix the issue, you can review it and merge to master.
[unity_internals.h.patch.txt](https://github.com/ThrowTheSwitch/Unity/files/2273065/unity_internals.h.patch.txt)

**Comment by mvandervoord** [comment]
I suspect you've misconfigured something. The cast you've replaced isn't something we can directly put into the code, because most embedded platforms don't support UINT64... but also it shouldn't be necessary. If you have everything configured properly, UNITY_INT is already an int64.

What do you have UNITY_LONG_WIDTH and UNITY_INT_WIDTH set to?

**Comment by alef75** [comment]
I'm sorry for misunderstanding..
I've not set them.
which is the correct value (and place) to set them?
My processor is an 32 bit Arm9:
int -->                 32 bit
long int -->         32 bit
long long int --> 64 bit

**Comment by mvandervoord** [comment]
No apology necessary. I'm in the process of writing an improved guide for this sort of configuration, for exactly this problem. It's a mess of tricky problems right now. There may still be bugs lurking too... hopefully not, but it's hard to hunt down all the combinations.

The current documentation is here: https://github.com/ThrowTheSwitch/Unity/blob/master/docs/UnityConfigurationGuide.md

The first section has a discussion on how Unity attempts to handle integer types. If it's still acting weird after tackling that, or something doesn't make sense, don't be afraid to say so. I'm not offended. ;)

**Comment by alef75** [comment]
I've already read the guide, and I read it again but I can't understand which values I have to set for my system. for UNITY_LONG_WIDTH UNITY_INT_WIDTH
in my system:
int --> 32 bit
long int --> 32 bit
long long int --> 64 bit

so I guessed the I need to set
UNITY_LONG_WIDTH = 32
UNITY_INT_WIDTH = 32

but with this configuration the the runner coreduped.
have I to set  
UNITY_LONG_WIDTH = 32 ??
UNITY_INT_WIDTH = 64 ??

**Comment by mvandervoord** [comment]
Hm. Excellent question. I would have expected 32 and 32 to get you there just fine. 

Before that, though, a coredump seems an extreme reaction to being cast to a smaller integer! I would think you'd just have the issue of it truncating your results.

Out of curiosity, what compiler are you using? Which assertion are you using when it coredumps?

Can you run an experiment real quick? If you instead had a test that was just `TEST_ASSERT_EQUAL_INT(4, 5)` does that fail without crashing?

**Comment by alef75** [comment]
The compiler is codesougery gcc 
arm-none-linux-gnueabi-gcc-4.8.3

the test cases use TEST_ASSERT_EQUAL_UINT64(1255555555555555555, data);
If I don't use TEST_ASSERT_XXXX_XINT64 all works right

Using my patch unity_internals.h.patch.txt, all works right.

**Comment by mvandervoord** [comment]
:) I believe you. I'm just not having much luck reproducing the problem and I'm perplexed why casting to a smaller int would cause this to crash. That seems like it doesn't even follow the C standard, though gcc is usually pretty good about this. I'm running a few varieties of gcc over here that all seem fine with it... so I'm trying to determine what is making yours act differently. 

If you add the following to a test, I strongly suspect that it will return `4 8`, even though it should be autodetecting to `8 8` 

```
printf("%i %i\r\n", sizeof(UNITY_INT), sizeof(UNITY_INT64));
```

Does it act any differently if you also add defines `UNITY_EXCLUDE_STDINT_H` and `UNITY_EXCLUDE_LIMITS_H` ?

**Comment by alef75** [comment]
Hi,
I'm wrong The problem was due to a misconfiguration.
I've reverted the original unity_internals.h and now all work right.
I've understood what was the issue:
In the make file for the test I was defining an old unity macro:
-DUNITY_INCLUDE_64 instead of   -DUNITY_SUPPORT_64
in this way I was obtaining that 64 bit tests fail (but in my mind it would work), so I've forced  
UNITY_LONG_WIDTH = 64
UNITY_INT_WIDTH = 64 
and this probably make things inconsistent.
Sorry for that
thank you for supporting me and for the great unity
